### ISSUE TITLE
MDRangePolicy Reduction requires value_type typedef in Functor

### ISSUE THREAD

**Issue by crtrott** [kokkos/kokkos#1379]
Apparently the value_type deduction is screwed up, and we do not have a test which uses either a Lambda or a functor without value_type typedef. 

Example:
```c++
Kokkos::parallel_reduce(Kokkos::MDRangePolicy<Kokkos::Rank<2>> ({0,0},{10,10}), KOKKOS_LAMBDA (const int i, const int j, int& lsum) {
    lsum += i*10+j;
  },result);
```

Error:
```
/home/crtrott/Kokkos/kokkos/core/src/Kokkos_Serial.hpp: In instantiation of ‘class Kokkos::Impl::ParallelReduce<main(int, char**)::<lambda(const int&, const int&, int&)>, Kokkos::MDRangePolicy<Kokkos::Rank<2u, (Kokkos::Iterate)0, (Kokkos::Iterate)0> >, Kokkos::InvalidType, Kokkos::Serial>’:
/home/crtrott/Kokkos/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:890:14:   required from ‘static void Kokkos::Impl::ParallelReduceAdaptor<PolicyType, FunctorType, ReturnType>::execute(const string&, const PolicyType&, const FunctorType&, ReturnType&) [with PolicyType = Kokkos::MDRangePolicy<Kokkos::Rank<2u, (Kokkos::Iterate)0, (Kokkos::Iterate)0> >; FunctorType = main(int, char**)::<lambda(const int&, const int&, int&)>; ReturnType = int; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/home/crtrott/Kokkos/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:978:74:   required from ‘void Kokkos::parallel_reduce(const PolicyType&, const FunctorType&, ReturnType&, typename Kokkos::Impl::enable_if<Kokkos::is_execution_policy<PolicyType>::value>::type*) [with PolicyType = Kokkos::MDRangePolicy<Kokkos::Rank<2u, (Kokkos::Iterate)0, (Kokkos::Iterate)0> >; FunctorType = main(int, char**)::<lambda(const int&, const int&, int&)>; ReturnType = int; typename Kokkos::Impl::enable_if<Kokkos::is_execution_policy<PolicyType>::value>::type = void]’
main.cpp:23:11:   required from here
/home/crtrott/Kokkos/kokkos/core/src/Kokkos_Serial.hpp:708:47: error: no type named ‘value_type’ in ‘Kokkos::Impl::ParallelReduce<main(int, char**)::<lambda(const int&, const int&, int&)>, Kokkos::MDRangePolicy<Kokkos::Rank<2u, (Kokkos::Iterate)0, (Kokkos::Iterate)0> >, Kokkos::InvalidType, Kokkos::Serial>::ReducerTypeFwd {aka struct main(int, char**)::<lambda(const int&, const int&, int&)>}’
   typedef typename ReducerTypeFwd::value_type ValueType; 
                                               ^
In file included from /home/crtrott/Kokkos/kokkos/core/src/Kokkos_Parallel.hpp:62:0,
                 from /home/crtrott/Kokkos/kokkos/core/src/Kokkos_Serial.hpp:55,
                 from /home/crtrott/Kokkos/kokkos/core/src/Kokkos_Core.hpp:53,
                 from main.cpp:1:
/home/crtrott/Kokkos/kokkos/core/src/impl/Kokkos_FunctorAnalysis.hpp: In instantiation of ‘struct Kokkos::Impl::FunctorAnalysis<Kokkos::Impl::FunctorPatternInterface::REDUCE, Kokkos::RangePolicy<Kokkos::Serial, Kokkos::Schedule<Kokkos::Static>, long unsigned int>, main(int, char**)::<lambda(const int&, const int&, int&)> >::deduce_value_type<main(int, char**)::<lambda(const int&, const int&, int&)>, Kokkos::Impl::FunctorPatternInterface::REDUCE, void, true>’:
/home/crtrott/Kokkos/kokkos/core/src/impl/Kokkos_FunctorAnalysis.hpp:206:70:   required from ‘struct Kokkos::Impl::FunctorAnalysis<Kokkos::Impl::FunctorPatternInterface::REDUCE, Kokkos::RangePolicy<Kokkos::Serial, Kokkos::Schedule<Kokkos::Static>, long unsigned int>, main(int, char**)::<lambda(const int&, const int&, int&)> >’
/home/crtrott/Kokkos/kokkos/core/src/Kokkos_Serial.hpp:714:46:   required from ‘class Kokkos::Impl::ParallelReduce<main(int, char**)::<lambda(const int&, const int&, int&)>, Kokkos::MDRangePolicy<Kokkos::Rank<2u, (Kokkos::Iterate)0, (Kokkos::Iterate)0> >, Kokkos::InvalidType, Kokkos::Serial>’
/home/crtrott/Kokkos/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:890:14:   required from ‘static void Kokkos::Impl::ParallelReduceAdaptor<PolicyType, FunctorType, ReturnType>::execute(const string&, const PolicyType&, const FunctorType&, ReturnType&) [with PolicyType = Kokkos::MDRangePolicy<Kokkos::Rank<2u, (Kokkos::Iterate)0, (Kokkos::Iterate)0> >; FunctorType = main(int, char**)::<lambda(const int&, const int&, int&)>; ReturnType = int; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/home/crtrott/Kokkos/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:978:74:   required from ‘void Kokkos::parallel_reduce(const PolicyType&, const FunctorType&, ReturnType&, typename Kokkos::Impl::enable_if<Kokkos::is_execution_policy<PolicyType>::value>::type*) [with PolicyType = Kokkos::MDRangePolicy<Kokkos::Rank<2u, (Kokkos::Iterate)0, (Kokkos::Iterate)0> >; FunctorType = main(int, char**)::<lambda(const int&, const int&, int&)>; ReturnType = int; typename Kokkos::Impl::enable_if<Kokkos::is_execution_policy<PolicyType>::value>::type = void]’
main.cpp:23:11:   required from here
/home/crtrott/Kokkos/kokkos/core/src/impl/Kokkos_FunctorAnalysis.hpp:163:34: error: no matching function for call to ‘Kokkos::Impl::FunctorAnalysis<Kokkos::Impl::FunctorPatternInterface::REDUCE, Kokkos::RangePolicy<Kokkos::Serial, Kokkos::Schedule<Kokkos::Static>, long unsigned int>, main(int, char**)::<lambda(const int&, const int&, int&)> >::deduce_value_type<main(int, char**)::<lambda(const int&, const int&, int&)>, Kokkos::Impl::FunctorPatternInterface::REDUCE, void, true>::deduce(void (main(int, char**)::<lambda(const int&, const int&, int&)>::*)(const int&, const int&, int&) const)’
     using type = decltype( deduce( & F::operator() ) );

```

**Comment by ndellingwood** [comment]
Made a similar test (with `double` as the result type)
`ReturnType` in `parallel_reduce` is `double` (as expected), but the `ReducerType` parameter in `ParallelReduce` is an `InvalidType` so I think what's happening is the `ValueType` Is trying to be set to the `value_type` of the non-existent functor (since a lambda was passed in).

 `ParallelReduceReturnValue<void, ReturnType, FunctorType>::reducer_type` coming back as an `InvalidType` rather than the `return_type` may be triggering this, still digging.

**Comment by ndellingwood** [comment]
Have a fix for this, having issues with my VPN and pushing the commit from home, will put in PR tomorrow.

**Comment by ndellingwood** [comment]
PR #1380 in for this issue.

**Comment by ndellingwood** [comment]
Related to #1054

**Comment by ndellingwood** [comment]
Fix in PR #1380 spot checked with test_all_sandia and should be ready for merge.
